Src/FluentAssertions/Equivalency/EquivalencyValidator.cs

```diff
@@ -44,87 +43,89 @@ public void AssertEquality(EquivalencyValidationContext context)
             }
         }
 
-        private bool IsComplexType(object @object)
+        public void RecursivelyAssertEquality(IEquivalencyValidationContext context)
         {
-            if (@object == null)
-                return false;
+            if (ShouldCompareMembersThisDeep(context.SelectedMemberPath))
+            {
+                UpdateScopeWithReportableContext(context);
 
-            Type type = @object.GetType();
+                if (!IsCyclicReference(context))
+                {
+                    RunStepsUntilEquivalencyIsProven(context);
+                }
+            }
+        }
+
+        private bool ShouldCompareMembersThisDeep(string selectedMemberPath)
+        {
+            const char memberSeparator = '.';
+            var depth = selectedMemberPath.Count(chr => chr == memberSeparator);
+            bool shouldRecurse = config.AllowInfiniteRecursion || depth < MaxDepth;
 
-            if (!isComplexTypeMap.TryGetValue(type, out bool isComplexType))
+            if (!shouldRecurse)
             {
-                isComplexType = !type.OverridesEquals();
-                isComplexTypeMap[type] = isComplexType;
+                AssertionScope.Current.FailWith("The maximum recursion depth was reached.  ");
             }
 
-            return isComplexType;
+            return shouldRecurse;
```

Comment:
`shouldRecurse` is always `true` at this point, so the function could return `void` instead.

Reply:
Unless somebody wrapped the call to `BeEquivalent` in another `AssertionScope`